### PR TITLE
Interactive transactions on Vax

### PR DESCRIPTION
**Pull request by v0idpwn**
Implements Ecto.Adapter.Transaction using :antidotec_pb interactive
transaction API. As all antidote calls must be done within a
transaction, a static transaction is opened if no explicit call
of `Repo.transaction/1` is made.Transaction nesting is allowed
in regular Ecto fashion.

### CODE REVIEW

File: apps/vax/test/integration/adapter_test.exs [hunk 1]
```diff
@@ -60,6 +60,47 @@ defmodule Vax.AdapterIntegrationTest do
...
+          assert_receive :finish_2
+        end)
+
+        send(parent, :transaction_2_finished)
+      end)
+
+    assert_receive :transaction_1_finished, 5000
+    assert_receive :transaction_2_finished, 5000
+
+    assert 30 = TestRepo.read_counter(id)
```
Comment (v0idpwn) [inline apps/vax/test/integration/adapter_test.exs]: When calling `antidotec_pb:commit_transaction/2`, for transaction 2, I get `{error, aborted}`. Transaction 1 commits as expected.

Comment (bieniusa) [reply apps/vax/test/integration/adapter_test.exs]: ClockSI requires that all transactions on the same key(s) at one DC are serialized. This is why the abort is triggered.

Comment (bieniusa) [reply apps/vax/test/integration/adapter_test.exs]: In the Antidote test suite, all txns are therefore restarted until they successfully commit.

**Comment by bieniusa** [comment]
I would find a commit message what type of txns, what the API is etc. very helpful.....

File: apps/vax/test/integration/adapter_test.exs (hunk 1 above)
Comment (v0idpwn) [reply apps/vax/test/integration/adapter_test.exs]: >ClockSI requires that all transactions on the same key(s) at one DC are serialized. This is why the abort is triggered.

Guess that alternatively we can disable this behaviour? Did that in https://github.com/vaxine-io/vaxine/pull/27/commits/93591ab8d89b5fceb35118b3e468b8457db55354 

cc @balegas

**Comment by v0idpwn** [comment]
>I would find a commit message what type of txns, what the API is etc. very helpful.....

I'm implementing this behaviour: https://hexdocs.pm/ecto/Ecto.Adapter.Transaction.html through interactive transactions.

When it is implemented, the user gets part of the Ecto API that was previously unavailable because the behaviour wasn't implemented, namely `Repo.transaction/1` and `Repo.rollback/1`.

I'll try to add docs/examples on readme/more tests, this is a bit of a wip

**Comment by balegas** [comment]
Agree with suggestions about docs.
Can you skip the test if certification is ON, or ensure it is off for the test? Vax should still be able to work with certification ON

**Comment by bieniusa** [comment]
Since turning it off can lead to problems, I rather suggest rewriting the test.

**Comment by v0idpwn** [comment]
But @bieniusa, I think rewriting the test is just ignoring the problem. This **should** be possible, shouldn't it? In my perspective, the transactions happening in the same or a different DC should be an irrelevant detail for the user. I understand the concern with reservations, but maybe we could verify only keys that really use reservations?

If this **shouldn't** be possible, then we should probably return a proper error, document why it happens and how one should handle this case.

**Pull request merged**